File: ios/Crew/Engine/SwapFinder.swift
// SPEC: Flow 1 step 4 (3–5 alternatives that do the same job) · 5.6.1 swapCandidates (same pattern, ≤5, never incumbent)
// · 8.3 · exercises.json swapRule (tiers swapGroup → pattern → region, widening only while fewer than swapCandidatesMin
// exist; same-or-lower level first). Twin of swap-finder.ts. WRITTEN — UNVERIFIED (needs Mac).

import Foundation

enum SwapFinder {
    private static let levels = ["brandNew", "some", "experienced"]
    private static var levelRank: [String: Int] { Dictionary(uniqueKeysWithValues: levels.enumerated().map { ($1, $0) }) }

    static func swapCandidates(for incumbent: SeedExercise, access: String, experience: String, seed: SeedCatalog) -> [SeedExercise] {
        let available = Set(seed.equipmentAccess[access] ?? [])
        let usable = seed.exercises.filter { $0.id != incumbent.id && $0.type == incumbent.type && available.contains($0.equipment) }
        let tiers: [(SeedExercise) -> Bool] = [
            { $0.swapGroup == incumbent.swapGroup },
            { $0.pattern == incumbent.pattern },
            { seed.regionOfPattern[$0.pattern] == seed.regionOfPattern[incumbent.pattern] },
        ]
        var found: [SeedExercise] = []
        for tier in tiers {
            found = usable.filter(tier)
            if found.count >= SpecConstants.swapCandidatesMin { break }
        }
        let userRank = levelRank[experience] ?? 0
        let ranked = found.sorted { left, right in
            let leftFits = (levelRank[left.level] ?? 0) <= userRank ? 0 : 1
            let rightFits = (levelRank[right.level] ?? 0) <= userRank ? 0 : 1
            if leftFits != rightFits { return leftFits < rightFits }
            let leftLevel = levelRank[left.level] ?? 0
            let rightLevel = levelRank[right.level] ?? 0
            if leftLevel != rightLevel { return leftLevel < rightLevel }
            return left.name < right.name
        }
        return Array(ranked.prefix(SpecConstants.swapCandidatesMax))
    }
}

File: web/src/lib/engine/swap-finder.ts
// SPEC: Flow 1 step 4 (Tap → Swap → 3–5 alternatives that do the same job; no questions asked) · 5.6.1 swapCandidates
// (same pattern, ≤5, never incumbent) · 8.3 (candidates share pattern + equipment) · exercises.json swapRule (tiers
// swapGroup → pattern → region, widening only while fewer than swapCandidatesMin exist; same-or-lower level ranked first).
// Twin: ios/Crew/Engine/SwapFinder.swift. Pure.
import type { EquipmentAccess, Experience, SeedExercise } from "@/generated/seed";
import { equipmentAccess, regionOfPattern } from "@/generated/seed";
import { SpecConstants } from "@/generated/spec-constants";

const LEVEL_RANK: Record<Experience, number> = { brandNew: 0, some: 1, experienced: 2 };

function rankForUser(experience: Experience) {
  return (left: SeedExercise, right: SeedExercise): number => {
    const leftFits = LEVEL_RANK[left.level] <= LEVEL_RANK[experience] ? 0 : 1;
    const rightFits = LEVEL_RANK[right.level] <= LEVEL_RANK[experience] ? 0 : 1;
    if (leftFits !== rightFits) return leftFits - rightFits;
    return LEVEL_RANK[left.level] - LEVEL_RANK[right.level] || left.name.localeCompare(right.name);
  };
}

export function swapCandidates(incumbent: SeedExercise, access: EquipmentAccess, experience: Experience, exercises: SeedExercise[]): SeedExercise[] {
  const available = new Set(equipmentAccess[access]);
  const usable = exercises.filter((candidate) => candidate.id !== incumbent.id && candidate.type === incumbent.type && available.has(candidate.equipment));
  const tiers = [
    (candidate: SeedExercise) => candidate.swapGroup === incumbent.swapGroup,
    (candidate: SeedExercise) => candidate.pattern === incumbent.pattern,
    (candidate: SeedExercise) => regionOfPattern[candidate.pattern] === regionOfPattern[incumbent.pattern],
  ];
  let found: SeedExercise[] = [];
  for (const tier of tiers) {
    found = usable.filter(tier);
    if (found.length >= SpecConstants.swapCandidatesMin) break;
  }
  return found.sort(rankForUser(experience)).slice(0, SpecConstants.swapCandidatesMax);
}
